### PR TITLE
introduces fantail unit tests

### PR DESCRIPTION
**Pull request by janvanrijn**

### CODE REVIEW

File: src/main/java/org/openml/webapplication/fantail/dc/statistical/SimpleMetaFeatures.java [hunk 1]
```diff
@@ -11,167 +11,142 @@
...
+			if (dataset.classAttribute().isNumeric()) {
+				TimeBasedChanges += Math.abs(previousInstance.classValue() - currentInstance.classValue());
+			} else if (dataset.classAttribute().isNominal()) {
+				TimeBasedChanges += (previousInstance.classValue() == currentInstance.classValue() ? 0 : 1);
+			}
+		}
+		AutoCorrelation = (NumberOfInstances > 1 ? new Double(NumberOfInstances - 1 - TimeBasedChanges) / new Double(NumberOfInstances - 1.0) : 1.0);
+
+		// Class properties
+		if (dataset.attribute(dataset.classIndex()).type() == Attribute.NUMERIC) {
```
Comment (joaquinvanschoren) [inline src/main/java/org/openml/webapplication/fantail/dc/statistical/SimpleMetaFeatures.java:100]: The characterizer assumes that the target variable is always known. This is not always the case, some datasets are on unsupervised problems, in others the target is not clear or simply unknown. The evaluation engine assumes that in those cases the last feature is always the target, but this is not always correct and leads to messy meta-features, e.g. see dataset 4133 which has a class count of 10104. We should probably not automatically assign a target class, but simply not compute supervised meta-features if the target is now known.

**Review by joaquinvanschoren — COMMENTED** [review]
Good, but we still get bad meta-features if the target feature is not specified by the user.

The characterizer assumes that the target variable is always known. This is not always the case, some datasets are on unsupervised problems, in others the target is not clear or simply unknown. The evaluation engine assumes that in those cases the last feature is always the target, but this is not always correct and leads to messy meta-features, e.g. see dataset 4133 which has a class count of 10104. We should probably not automatically assign a target class, but simply not compute supervised meta-features if the target is now known.

**Comment by janvanrijn** [comment]
> The characterizer assumes that the target variable is always known. This is not always the case, some datasets are on unsupervised problems, in others the target is not clear or simply unknown. The evaluation engine assumes that in those cases the last feature is always the target, but this is not always correct and leads to messy meta-features, e.g. see dataset 4133 which has a class count of 10104. We should probably not automatically assign a target class, but simply not compute supervised meta-features if the target is now known.

Yes, but that is a completely different topic (and an incredibly difficult one, as for example the recently introduced multitask datasets are badly defined and unclear how to handle these)

This PR focuses on improving the code quality of the evaluation engine and unit testing all the Characterizer classes. It would be great if you could review this PR within this scope.

**Review by joaquinvanschoren — APPROVED** [review]
Looks fine. One more nagging doubt is whether we need to emit null qualities (e.g. landmarker_AUC = null for regression datasets). Probably you want to store them and have it on record that we tried computing them but couldn't? if so, just ignore this and merge.

**Comment by janvanrijn** [comment]
> Probably you want to store them and have it on record that we tried computing them but couldn't?

This is exactly what we do by returning null. OpenML will store this as such which means: tried but was not applicable. we should document this somewhere

**Pull request merged**